### PR TITLE
identity: version check multiple and implicit identities

### PR DESCRIPTION
**Pull request by tgross**
Job submitters cannot set multiple identities prior to Nomad 1.7, and cluster administrators should not set the identity configurations for their `consul` and `vault` configuration blocks until all servers have been upgraded. Validate these cases during job submission so as to prevent state store corruption when jobs are submitting in the middle of a cluster upgrade.

### CODE REVIEW

File: nomad/job_endpoint_hooks.go [hunk 1]
```diff
@@ -402,7 +409,19 @@ func (v *jobValidate) Validate(job *structs.Job) (warnings []error, err error) {
 	return warnings, validationErrors.ErrorOrNil()
 }
 
-func (v *jobValidate) validateServiceIdentity(s *structs.Service, parent string) error {
+func (v *jobValidate) isEligibleForMultiIdentity() bool {
+	if v.srv == nil || v.srv.serf == nil {
+		return true // handle tests w/o real servers safely
+	}
+	return ServersMeetMinimumVersion(
+		v.srv.Members(), v.srv.Region(), minVersionMultiIdentities, true)
```
Comment (schmichael) [inline nomad/job_endpoint_hooks.go:417]: Would be nice if ServersMeetMinimumVersion was a struct that wrapped up `Members()` and `Region()` and only provided a simple `Meets(version, checkFailed)` interface for us to mock/stub out in tests.

nbd though, nothing to block this work on.

File: nomad/job_endpoint_hooks.go [hunk 2]
```diff
@@ -402,7 +409,19 @@ func (v *jobValidate) Validate(job *structs.Job) (warnings []error, err error) {
...
+}
+
+func (v *jobValidate) validateServiceIdentity(s *structs.Service, parent string, okForIdentity bool) error {
+	if s.Identity != nil && !okForIdentity {
+		return fmt.Errorf("Service %s in %s cannot have an identity until all servers are upgraded to %s",
```
Comment (schmichael) [inline nomad/job_endpoint_hooks.go:422]: Just a little pedantic nitpicking on the wording.

```suggestion
		return fmt.Errorf("Service %s in %s cannot have an identity until all servers are upgraded to >= %s",
```

or

```suggestion
		return fmt.Errorf("Service %s in %s cannot have an identity until all servers are upgraded to %s or later",
```

**Review by schmichael — APPROVED** [review]
Thanks. I think this is a great way to protect against truly wacky behavior in mixed version settings.

An implicit constraint on the client being >= 1.7.0 would be useful too.

File: nomad/job_endpoint_hooks.go (hunk 1 above)
Comment (tgross) [reply nomad/job_endpoint_hooks.go:417]: I've opened https://github.com/hashicorp/nomad/issues/18931 to follow-up on that when I have a few minutes.

**Comment by tgross** [comment]
> An implicit constraint on the client being >= 1.7.0 would be useful too.

Agreed. I'll follow-up with that in a separate PR.

**Comment by tgross** [comment]
Follow-up for the client version check in https://github.com/hashicorp/nomad/pull/18932

**Pull request merged**